### DISCUSSION TITLE
[DEVICE_ADD] POW-HVM6.2K-48V-LIP Add Modbus support

### DISCUSSION THREAD

**Discussion by v-moto** [softwarecrash/Solar2MQTT-ESP8266#209] [Q&A]
### Is your feature request related to a problem? Please describe

I cannot connect a new inverter to the HA

### Describe the solution you'd like

I received a description of the inverter's communication protocol from the manufacturer, please add minimal support for this device
[Modbus_RTU communication description.pdf](https://github.com/user-attachments/files/17908944/Modbus_RTU.communication.description.pdf)

### Describe alternatives you've considered

I did not find alternatives for this device, and it is not advisable to buy their adapter because it cannot be connected to HA

**Comment by softwarecrash** [comment]
Let it Check later.
@bigdigital See you a Chance to build it in?

**Comment by v-moto** [comment]
@softwarecrash @bigdigital 
Do you have a usdt trc20 wallet for donations? Because I really like your project and want to support you a little) and maybe it will speed up the adaptation of my inverter a little))

**Comment by softwarecrash** [comment]
Badly No, PayPal ist linked in the Project for donation.
But sorry it Doenst Speed it Up, i have much Work on buissenes and less free time. So when i can get a Bit time i will Work on it.
Its Bit Tricky to Work without the device, i will try it when find some time

**Comment by v-moto** [comment]
Okay.
Then I'll send it via Paypal one of these days. 
Thank you very much)
If you need any help with the tests, I am always happy to help

**Comment by bigdigital** [comment]
@v-moto pls contact me in discord so i can send you a test version

**Comment by sMusia** [comment]
Hi, I’m also an owner of such an inverter. I can join the tests too!

**Comment by v-moto** [comment]
> Hi, I’m also an owner of such an inverter. I can join the tests too!

I have already received the beta, today I will test it, if everything goes well, I will write the result

**Comment by v-moto** [comment]
> Hi, I’m also an owner of such an inverter. I can join the tests too!

Where are you from? And do you have an original adapter for this inverter?

**Comment by sMusia** [comment]
> > Hi, I’m also an owner of such an inverter. I can join the tests too!
> 
> Where are you from? And do you have an original adapter for this inverter?

I am from Ukraine. Unfortunately, I don’t have the original adapter. Here's the story: I had a POW-HVM-3.5H-24V inverter. At that time, there was no information on the internet about the communication protocol with the inverter, so I independently developed a way to retrieve some telemetry for HomeAssistant. Later, I switched to ESPHOME, which provided more telemetry data. Now, I’ve upgraded to the POW-HVM6.2K-48V-LIP and discovered the com port (RS232) and RJ45 (RS485). I couldn’t establish communication with either port—the inverter simply didn’t respond. That’s when I started thinking that the folks at POWMR might have changed the communication protocol. Then I came across Solar2MQTT and this query...

**Reply by r3ndszergazda** [reply]
How did you communicate with the POW-HVM-3.5H-24V ?

**Reply by sMusia** [reply]
Solar2MQTT

**Comment by sMusia** [comment]
Figured out the registers and made a working version based on ESPHOME.
![image](https://github.com/user-attachments/assets/3ef577ac-6470-42e9-b58e-cf380feada3a)

**Comment by essergey** [comment]
@sMusia How did you manage to get the data?

**Comment by tsdaemon** [comment]
> Figured out the registers and made a working version based on ESPHOME. ![image](https://private-user-images.githubusercontent.com/40174981/396062948-3ef577ac-6470-42e9-b58e-cf380feada3a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1Nzk0NjgsIm5iZiI6MTczNTU3OTE2OCwicGF0aCI6Ii80MDE3NDk4MS8zOTYwNjI5NDgtM2VmNTc3YWMtNjQ3MC00MmU5LWI1OGUtY2YzODBmZWFkYTNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjMwVDE3MTkyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiMDJkM2YzZjEyMGExN2VlYmY4NjNkN2VmMzQ0MWM0ZTE5YjE1OGI0MmEwYmQ3ZGIxZTM0Y2EyOTIxMjVjYjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.CpbiRsw4u_5R6C05VsQdx7rms2Wtc9vG4lO2OBEUWg0)

Hi! Thank you for sharing, can you share more details how did you do this with ESPHOME? I am also from Ukraine, and I am trying to set the monitoring for the same inverter.

**Comment by sMusia** [comment]
> > Figured out the registers and made a working version based on ESPHOME.
> 
> Hi! Thank you for sharing, can you share more details how did you do this with ESPHOME? I am also from Ukraine, and I am trying to set the monitoring for the same inverter.

[inverter-powmr.zip](https://github.com/user-attachments/files/18280365/inverter-powmr.zip)


I used ESP8266, connection: esp: tx_pin: GPIO1, rx_pin: GPIO3 -> (RS-232 to UART, MAX3232) -> com_port: pin 2 - TX, pin 3 - RX, pin 5 - GND, pin 9 - +12-15V (with a step-down power module for ESP8266 and MAX3232). 

ESPHOME yaml in attachment.

Very important!: Make as few requests as possible on the Modbus bus, refer to - Optimizing modbus communications - https://esphome.io/components/modbus_controller

**Comment by gyp5y** [comment]
Hi there!
I've just got data from my POW-HVM6.2K-48V-LIP using sample provided by @sMusia . It's great! Thank you!
The only thing that looks strange it's "Working Mode" values. I've got 8 and 9 on mine. Do you have extended description for those modes by any chance?

**Reply by Massterzinger** [reply]
Hi, have you figured it out?
Also I have wrong values in output power, it shows around 1000w when should show ~1700w

**Reply by gyp5y** [reply]
Hi. Yeah. It's fixed now. Currents there was 10x multiplied so it's 80 and 15 (that was also fixed). Check my ESPHome config https://github.com/user-attachments/files/18574441/powmr_lip.yaml.txt.

**Reply by Massterzinger** [reply]
Oh, my bad, installed the wrong file, thanks for the config.
Still have working mode 08 and Output power shows wrong

**Reply by gyp5y** [reply]
It looks like it wasn't the last version. Attached current config. See below.

**Reply by Massterzinger** [reply]
That's it, thank you, it works great now.
BTW I have 2 inverters in parallel, probably that's why it shows around a half of production in the Active Power)

**Reply by gyp5y** [reply]
Maybe. I have one for now so I can't check how it's working with paralleled inverters or three-phase scheme.

**Reply by gyp5y** [reply]
Btw you can add this code to see (or even switch) output mode. It's removed in original yaml
``` 
- platform: modbus_controller
    modbus_controller_id: smg_inverter
    name: "Output mode"
    entity_category: config
    address: 300
    value_type: U_WORD
    use_write_multiple: true
    optionsmap:
      "Single": 0
      "Parallel": 1
      "3 Phase P1": 2
      "3 Phase P2": 3
      "3 Phase P3": 4
```

**Reply by Massterzinger** [reply]
Wow, thank you.
Btw, does your inverter also charge the battery from utility only in USB (utility-pv-battery) mode?

**Reply by gyp5y** [reply]
I have no PV for now. It looks so

**Reply by Massterzinger** [reply]
Figured it out, when selecting PV-Utility-Battery, inverter should set into SUB setting, but has SOL instead. Probably something wrong with command

**Reply by gyp5y** [reply]
I switch mine between PV-Utility-Battery and Utility-PV-Battery to relocate grid usage on cheaper night tariff. It works just fine. Could you describe your case in detail? Btw what charging priority is set?

**Reply by Massterzinger** [reply]
It's strange, Inverter has 4 options in output source priority, but in the PDF only 3 are described.
Inverter has:
SOL, Uti, SBU, SUB.
PDF and yaml doesn't mention SOL, but sets to it when SUB (PV-Uti-Batt) is selected.
When I select SUB on inverter manually - it works as expected, charging the battery from both PV and Utility.
Basically when you get PV, you should only switch from SBU (Day) and SUB (Night)

**Reply by gyp5y** [reply]
Are you sure that you don't mix up "Charging priority" with "Output priority"? I'll check both on the inverter later today.

**Reply by Massterzinger** [reply]
![20250219_113136](https://github.com/user-attachments/assets/eac47566-b731-4897-bf2d-6b12680a590b)
![20250219_113100](https://github.com/user-attachments/assets/c5f19019-fa1a-462b-8810-13db5d7b42c2)
Here, you can see both in user manual for the inverter

**Reply by Massterzinger** [reply]
Figured it out, that's how it should be

```
- platform: modbus_controller
    modbus_controller_id: smg_inverter
    name: "Output priority"
    entity_category: config
    address: 301
    value_type: U_WORD
    use_write_multiple: true
    optionsmap:
      "Utility-Only": 0
      "PV-Only": 1
      "PV-Battery-Utility": 2
      "PV-Utility-Battery" : 3
```

**Reply by gyp5y** [reply]
I didn't check all the settings from original yaml so it could be. Thanks. I'll adjust mine.

**Reply by gyp5y** [reply]
Hi. Do you know what does "Remote switch" (ID: 420) do? It was on for my inverter and after some time it shut off and I can't change it back.

**Reply by Massterzinger** [reply]
Hi, I was assuming it is what it says, something like software equivalent of the switch button on the side of the inverter.
BTW, have you ever had inverter send ~65535w of output active power
![image](https://github.com/user-attachments/assets/7fe0807e-92da-48ae-8835-132b49b2df65)

**Reply by gyp5y** [reply]
Indeed. Looks like there's a mess with register_count. According to pdf 206,209,213,217,220,224 have one register size but from original yml there are some random numbers set. Commented it out. If all will be okay I'll put updated config here.
I've got some kind of glitch few days age when inverter automagically switched to "StandBy" mode and turned output off without any errors. I was hoping to find explanation in values or statuses but no luck. "Remote switch" turned inverter on but left on "off" state. Still don't understand how it should work

**Reply by Massterzinger** [reply]
Chinese magic)

**Reply by Massterzinger** [reply]
Do you also have earth and neutral connected?

**Reply by gyp5y** [reply]
Yes. Both connected

**Reply by gyp5y** [reply]
It looks that those spikes up to int size is not caused by wrong register_count. Still has some.

**Reply by Massterzinger** [reply]
Maybe it is signed value, and it wants to pass something like -80w

**Reply by gyp5y** [reply]
You are right! There's a lot of values that are set to U_WORD in original yaml but they should be signed according to comm protocol.
Here's mine current config with signed/unsigned fixed.
[inverter-powmr-esp32.yaml.txt](https://github.com/user-attachments/files/19272636/inverter-powmr-esp32.yaml.txt)

**Reply by gyp5y** [reply]
I've figured out why switch component worked strange. Every switch should have "bitmask 1" option set 'cos it expects to have 0x01 value as on. Now Remote switch, LCD Backlight and others are working fine. Here's updated config.
[inverter-powmr-esp32.yaml.txt](https://github.com/user-attachments/files/19322016/inverter-powmr-esp32.yaml.txt)

**Comment by Fazabein** [comment]
Hello, my friend, can you make a tutorial on how to make a BMS cable? and which one should lead to the battery and which one should lead to the inverter? I use a hybrid inverter POW-HVM6.2K-48V-LIP and 2 amsan 48v 100ah wall mounted batteries 

![Image](https://github.com/user-attachments/assets/4d5a5a23-92e3-4349-bae7-418583741751)

. Can you help me ??

**Comment by gyp5y** [comment]
I use pin-to-pin straight RJ-45 cable connected to RS-485 port on BMS and inverter but your case could be different if your batteries aren't based on JK-BMS hardware. Then I was playing with protocol selection on BMS and on inverter. I don't remember  which one I'm using. I will check in a few days and write it here.

**Reply by gyp5y** [reply]
Protocol on the JK-BMS is (9600)PYLON_low_voltage and on the inverter is LI2 (PYLON US2000 according to manual)

**Comment by gyp5y** [comment]
@sMusia Using your example makes it read okay but I can't change any value. Maybe you can suggest how to solve this?
[12:34:58][D][select:062]: 'Battery charging priority' - Setting
[12:34:58][D][select:115]: 'Battery charging priority' - Set selected option to: Only PV charging
[12:34:58][D][modbus_controller.select:048]: Found value 3 for option 'Only PV charging'
[12:34:59][W][modbus_controller:027]: Modbus device=1 set offline
[12:34:59][D][modbus_controller:040]: Modbus command to device=1 register=0x14B no response received - removed from send queue
[12:35:00][W][modbus_controller:065]: Modbus device=1 back online
[12:35:00][D][modbus_controller.select:014]: New select value 0 from payload

**Reply by gyp5y** [reply]
This problem was solved using use_write_multiple: true for switches and other output sensors. PowMR supports only 0x10 function code when setting values. By default modbus controller in esphome sends 0x06 function code. use_write_multiple force it to use supported function code. Hope this will help someone. 
My config attached
[powmr_lip.yaml.txt](https://github.com/user-attachments/files/18549652/powmr_lip.yaml.txt)

**Reply by gyp5y** [reply]
Here's version with some fixes (register sizes, measurement units, max current step and display etc)
[powmr_lip.yaml.txt](https://github.com/user-attachments/files/18574441/powmr_lip.yaml.txt)

**Reply by gyp5y** [reply]
Here's one more update. It's version I use for a while and it's pretty fine.
[powmr_lip.yaml (1).txt](https://github.com/user-attachments/files/18824635/powmr_lip.yaml.1.txt)

**Reply by Massterzinger** [reply]
[inverter-powmr-esp32.yaml.txt](https://github.com/user-attachments/files/18866853/inverter-powmr-esp32.yaml.txt)
Here's my version for esp32 with fixed output source priority

**Comment by gyp5y** [comment]
If, by any chance, someone knows register address to get data equivalent to menu options 43,44,45 (SOC back-to-battery, SOC back-to-utility, SOC cut-off) - please share.

**Comment by synthesis77** [comment]
Thanks to those pioneering this!  I just bought three of these POW-HVM6.2K-48V-LIP units to replace a combination of Victor NM-II 5KW units and an EASUN ISolar-SMH-II-7KW-Wifi unit.  To my surprise, the WiFi-Pro dongles that were working with the older units don't establish coms with the new unit (though they power up and connect to wifi still).

So, since Smart ESS sucks anyway, I'm headed down the path of integrating ESPs into my setup.  I'm handy coding actual arduino /c code and was planning on rolling my own, but a prebuilt like this certainly makes things easier.

On a side note, I just got my RS232-TTL adapters (with built in male DB9s) and I just want to verify pin numbers.  I have this as a pinout, and am curious about the DB9 pins 2&3 possibly being reversed?
![51gcBY5Ym4L _AC_UF1000,1000_QL80_](https://github.com/user-attachments/assets/1ddc6e0a-2968-4bc5-ba52-f1fb25cc3522)

**Reply by gyp5y** [reply]
Hello, I'm sure that RX and TX are pins 2 and 3 but I not sure if they are reversed. I was playing with DB-9 to RJ-45 adapter and with RX, TX pins in ESP32 firmware so even if they are reversed it was fixed in firmware :) Inverter is pretty far from me so I can't check it for you. Pins 5 and 9 are VCC and GND. Voltage on VCC pin is smth about 15V so be sure to use voltage converter before powering your MCU using VCC from inverter.

**Reply by synthesis77** [reply]
as an FYI, your code worked GREAT!  I'm tuning it a bit and will upload it back, and then start to work on the parallel part, aggregating multiple POW-HVM6.2K-48V-LIPs into one virtual inverter bank.  Thank you!!

**Reply by gyp5y** [reply]
I'm glad that it's useful :) I have no additional inverters to test parallel functions so I skipped this part.

**Comment by gyp5y** [comment]
At last! I've got addresses not documented in protocol specification from PowMr support team so options 43, 44 and 45 are controllable too.
Here's updated config I'm using.
[inverter-powmr-esp32.yaml.txt](https://github.com/user-attachments/files/19936915/inverter-powmr-esp32.yaml.txt)
P.S. There is commented section to control battery discharge for battery communicating with inverter and not communicating. Comment out one you're not using. Or just uncomment both and keep in mind that only one section is used

**Comment by synthesis77** [comment]
Okay, thanks to you folks, I have parallel inverters up and going on a single ESP32.  I did a MAJOR refactor of the code you provided, and while it's sorta more complicated, it's more flexible and hopefully easier to use.

There's now TWO files (though you could use just the single one if you only have one inverter).  For me, it's inverter.yaml and single-pow-hvm62k-48v-lip.yaml.  I use the Packages feature to aggregate the "single" file into the top inverter file, and then create some prefixed aggregate sensors which combine the multiple inverters data into one.

If you only use the single file, you'll be just fine too, and if you ever update to multiple units, your data will be in the same place (i.e. the aggregate IDs will match the previous single IDs).  Note that I changed most of the data IDs though, so they won't line up with your previous data.  I also changed most of the names so that they'd be in groups, like "Output", "Inverter", "Battery", etc, so they sort better in the default HA UI.

I fixed several of the sensors' issues, ranging from missing units, types, etc, and tried to optimize it with register counts to make the data transmission more efficient.  I have it updating every second, and it works great even with two inverters.  There are some remaining issues (marked FIXME) like incorrect ranges on the battery settings, missing EQ settings, and the warnings/faults parameters weren't right (they're bitmasks, not values) so I'll clean those up too.  I also want to try to use anchors/aliases to reduce the amount of copy/pasting.

I should probably put these up in github, unless you want to [sMusia](https://github.com/sMusia) or [gyp5y](https://github.com/gyp5y)!

[inverter.yaml.txt](https://github.com/user-attachments/files/20173649/inverter.yaml.txt)
[single-pow-hvm62k-48v-lip.yaml.txt](https://github.com/user-attachments/files/20173667/single-pow-hvm62k-48v-lip.yaml.txt)

**Reply by gyp5y** [reply]
Great job! I'll definitely give it a try in a few days

**Comment by synthesis77** [comment]
I've updated the yaml to be totally anchored/aliased which makes it like 3 times smaller.  it also makes it so you can easily add more parameters and/or use it for other stuff.  I added the rest of the parameters too, and corrected most things.  still to do is fix the warning and fault flags from selects, but I can't decide if I want to expose that as a state string or as a number or as a ton of readonly coils/bitmasks...

sensors now look like this:
```
  - <<: *modbus_volts
    name: "Battery Voltage${name_suffix}"
    id: Battery_Voltage${id_suffix}
    address: 215
```

instead of this:
```
  - platform: modbus_controller
    modbus_controller_id: modbus_controller_id${id_suffix}
    name: "Battery Voltage${name_suffix}"
    id: Battery_Voltage${id_suffix}
    address: 215
    register_type: holding
    value_type: S_WORD
    state_class: measurement
    device_class: voltage
    unit_of_measurement: "V"
    accuracy_decimals: 1
    filters:
      - multiply: 0.1
```

it's now up in github:
https://github.com/synthesis77/esp32-stuff/tree/main

**Comment by vmomryk** [comment]
Hi there!
I'm using a PowMr 6.2KW 4.2KW Hybrid Solar Inverter in my home power supply system.
I've decided to integrate it to Home Assistant for the whole process automation.
I've made a wifi adapter based on Wemos D1 mini board using the [video manual](https://www.youtube.com/watch?v=ITDI0wnn0vE) published at PowMr official Youtube channel. Flashed the adapter using this project FW.
It works fine, all the data from inverter is being received by Home Assistant server, but unfortunately the adapter which is connected to RS232 port is permanently losing the connection and becoming unavailable.
So I'm trying to figure out how to fix this issue.
Is there 12v permanently on inverter rs232 pin 4? Maybe wifi adapter is switching off?
Or I need to use some kind of serial port isolator to avoid any electrostatic issues?
Can you advice me anything?
<img width="1039" height="588" alt="image (2)" src="https://github.com/user-attachments/assets/0215e491-3130-4f55-9bba-508f7a8dafd9" />

**Reply by synthesis77** [reply]
A few things to try here:  1st, it might be easier to just plug in a standard USB power brick to your outlet to try.  2nd, I have this rebooting problem too, and I think I’ve narrowed it down to when the host HAOS runs out of disk space.  There’s an API (and wifi) Timeout that forcibly reboots the device if the host has issues.  Pretty lame if you ask me, like, HAOS should cap its own db and not hang half of itself, and then hang connected devices.  You could try api: -timeout:0 to disable it.  3rd, if you shut down the inverter, the 12v rail shuts off too.  4th, you can check the logs of the device in ESPHome for more info.

ps.  I should update my newest github link, I’ve made a few more updates

**Reply by vmomryk** [reply]
Thank you for a reply. Actually, I'm using inverter pin4 for a 12v, so I have to cut it off to being able to use 5v from USB.
Also, the point is that inverter should be always ON so 12v is always available. I am trying to avoid using any additional power suppliers for Wemos. I want to have it all as minimalistic as possible. Not sure I got you about HAOS settings, but will try to google it. 
UPD: I've just changed wifi channels selection from Auto to manual and it has been working whole day. I hope it will work tomorrow as well ;-))
UPD: UPD: unfortunately it failed after 5 hours of working fine ((

**Reply by synthesis77** [reply]
did you ever figure out what was going on?  my suggestion on usb power was not a permanent solution, it was just a debugging step to rule things out.  understood if it's hardwired though!

as a debugging tool, I added a simple blinking light to make sure the device isn't dead:

``` # LED BLINKING
output:
  - platform: gpio
    pin:
      number: 8
      mode: output
    id: led_id
interval:
  - interval: 1000ms
    then:
      - output.turn_on: led_id
      - delay: 250ms
      - output.turn_off: led_id
```

you could also add this, and then you can connect *directly* to your esp32, view its sensors and logs, change its settings, all without your Home Assistant even alive:

```
web_server:
  local: true
  version: 3 # <<< this line is optional, it might look nicer
```

and if the timeout is your problem (aka, the device is still alive, reachable via http://yourdevicename.local but it's unavailable in HA) you could add `-timeout:0` to your `api:` yaml section.  I haven't done it yet, but I will soon.

**Comment by gyp5y** [comment]
@synthesis77 Hello. It looks like now it's my turn to deal with two parallel inverters ) Could you answer me some questions? Is there a way to get data from slave inverter via adapter connected to master or they are not on the same bus? How you control e.g. charging current it seems to be x2 now when I set it on master and all other data seems to be split too.

**Reply by synthesis77** [reply]
hey, sorry I didn't reply.

> data from slave inverter via adapter connected to master

unfortunately, no.  you need a separate rs232 connection to each device to RECEIVE data from each inverter.  for TWO inverters, you can still do it with one esp32, just connect two TTL/RS232 adapters to it.  if you have more (I have 4) and exceed the number of UARTs, you can use multiple esp32s and use the packet_transport protocol to sync them together (as in my inverter.yaml file).


> How you control e.g. charging current

this in unfortunate in the exact OPPOSITE way: any setting set on one unit gets forcibly set on all other units via its own "parallel cable protocol" (which is likely just spewed commands without device IDs over 485 or something).

what this means in practice, is that you can't set them individually (even with the on-device buttons!)  so if you set 50A max charging, you'll get a max of 100A because each of your two inverters will do 50.  worse, if one is shaded and can only provide 20A, and one *could* do 80A, you won't get 100A total, you'll only get 70A (20A + 50A max).

I'll update my GitHub in a bit with newer files (I've made a bunch of changes) but the existing ones should work too.
